File: DevEnvironment/DataStructures/src/main/java/datastructures/bitarray/core/BitArrayReverseIterator.java
package datastructures.bitarray.core;

import base.core.Conditions;
import base.interfaces.IReverseIterator;
import datastructures.bitarray.interfaces.IBitArray;

/**
 * The BitArrayReverseIterator class implements a reverse iterator of a bit array.
 */
public final class BitArrayReverseIterator implements IReverseIterator<Boolean> {
    private final IBitArray bitArray;
    private int currIndex;

    /**
     * Creates a reverse iterator of a bit array.
     */
    public static IReverseIterator<Boolean> make(IBitArray bitArray) {
        return new BitArrayReverseIterator(bitArray);
    }

    /**
     * The BitArrayReverseIterator constructor.
     */
    private BitArrayReverseIterator(IBitArray bitArray) {
        Conditions.validateNotNull(
            bitArray,
            "The bit array.");

        this.bitArray = bitArray;

        this.reset();
    }

    /**
     * Checks whether there is a next element.
     */
    @Override
    public boolean hasNext() {
        return this.currIndex >= 0;
    }

    /**
     * Gets the next element.
     */
    @Override
    public Boolean next() {
        assert(this.hasNext());

        boolean currElement = this.bitArray.isOn(this.currIndex);
        --this.currIndex;

        return currElement;
    }

    /**
     * Resets the iterator.
     */
    @Override
    public void reset() {
        this.currIndex = this.bitArray.size() - 1;
    }
}

File: DevEnvironmentCLion/DataStructures/bitarray/core/BitArrayReverseIterator.cpp
#include "PreCompiled.h"
#include "BitArrayReverseIterator.h"
#include "IBitArray.h"

using namespace datastructures::bitarray;

/**
 * Creates a reverse iterator of a bit array.
 */
base::IReverseIteratorSharedPtr<bool> BitArrayReverseIterator::make(const IBitArray& bitArray)
{
    return std::make_shared<BitArrayReverseIterator>(bitArray);
}

/**
 * The BitArrayReverseIterator constructor.
 */
BitArrayReverseIterator::BitArrayReverseIterator(const IBitArray& bitArray) :
    m_bitArray(bitArray)
{
    reset();
}

/**
 * The BitArrayReverseIterator destructor.
 */
BitArrayReverseIterator::~BitArrayReverseIterator()
{
}

/**
 * Checks whether there is a next element.
 */
bool BitArrayReverseIterator::hasNext() const
{
    return !m_end;
}

/**
 * Gets the next element.
 */
bool BitArrayReverseIterator::next()
{
    assert(hasNext());

    bool currElement = m_bitArray.isOn(m_currIndex);

    if (m_currIndex == 0)
    {
        m_end = true;
    }
    else
    {
        --m_currIndex;
    }

    return currElement;
}

/**
 * Resets the iterator.
 */
void BitArrayReverseIterator::reset()
{
    m_currIndex = m_bitArray.size() - 1;
    m_end = false;
}
